### ISSUE TITLE
Don't use KeyboardEvent.code when resolving user keystrokes

### ISSUE THREAD

**Issue by Ben3eeE** [atom/atom-keymap#222]
## Description

It would be great if we could remove resolving keystrokes using `KeyboardEvent.code` because the code is the physical position of the key which doesn't change when a user remaps keys for example by using `xmodmap`. There is also a report where `KeyboardEvent.code` is incorrect but `KeyboardEvent.key` is correct. Likely because of a bug in Chrome.

[MDN](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code) says this about `KeyboardEvent.code`:
> The KeyboardEvent.code property represents a physical key on the keyboard (as opposed to the character generated by pressing the key). In other words, this property returns a value which isn't altered by keyboard layout or the state of the modifier keys.
This property is useful when you want to handle keys based on their physical positions on the input device rather than the characters associated with those keys

Currently all reports where this is a problem are for Backspace:

* There is a regression in Atom 1.12.0-beta2 when we started resolving backspace using `KeyboardEvent.code` that makes the E key resolve to backspace when using x2go or XQuartz. This is likely due to a bug in Chrome.

* It is quite common for users of the colemak layout to change positions of Tab and Backspace using for example `xmodmap`. This causes us to resolve both Tab and Backspace to Backspace. This is also a 1.12.0-beta2 regression.

Resolving using `KeyboardEvent.key` would fix both of these issues without the need for users to work around this using the [public keystroke resolver API](https://github.com/atom/atom-keymap/pull/164)

## Where

`atom-keymap` has two places where keystrokes are resolved using `KeyboardEvent.code`.

#### In `KEY_NAMES_BY_KEYBOARD_EVENT_CODE`

https://github.com/atom/atom-keymap/blob/3b17ede7f10e5a9c5097b495a03e88e2c7bacfa3/src/helpers.coffee#L7-L10

`Backspace` was added [to work around a bug](https://github.com/atom/atom-keymap/pull/154) with `ctrl-backspace` on Windows.

`Space` was added because `KeyboardEvent.key` [is a space character](https://github.com/atom/atom-keymap/pull/150) which we don't want to resolve to.

#### In `NUMPAD_KEY_NAMES_BY_KEYBOARD_EVENT_CODE`

https://github.com/atom/atom-keymap/blob/3b17ede7f10e5a9c5097b495a03e88e2c7bacfa3/src/helpers.coffee#L19-L30

This was added to [resolve the numpad digits differently](https://github.com/atom/atom-keymap/pull/198) to avoid changing tabs when trying to enter unicode characters using alt-numpad entry on Windows.

#### Getting rid of `KEY_NAMES_BY_KEYBOARD_EVENT_CODE`

1. Investigate resolving `Space` by checking if `KeyboardEvent.key` is a space character instead of checking if `KeyboardEvent.code` is `Space`. It could maybe be added to `NON_CHARACTER_KEY_NAMES_BY_KEYBOARD_EVENT_KEY`.

https://github.com/atom/atom-keymap/blob/3b17ede7f10e5a9c5097b495a03e88e2c7bacfa3/src/helpers.coffee#L11-L18

2. Investigate if the `Backspace` workaround is still needed. [Chrome has since fixed a bug](https://bugs.chromium.org/p/chromium/issues/detail?id=6941) where it didn't mask off the event code correctly when `Num Lock` or `Caps Locks` was on. I wonder if there are similar fixes that have reached us so we can remove this safely. I have tested this and it looks like the workaround is no longer needed.

#### Getting rid of `NUMPAD_KEY_NAMES_BY_KEYBOARD_EVENT_CODE`

Instead of checking if `KeyboardEvent.code` is `Numpad#` we could check if `KeyboardEvent.key` is a digit and use [KeyboardEvent.location](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/location) to identify if the digit comes from the numpad. `KeyboardEvent.location` is `3` for the numpad.

This would require checking it the location changes accordingly when remapping the numpad.

**Comment by nathansobo** [comment]
@Ben3eeE Do you want to give this a try? All the nuances of keyboard handling are hard to keep loaded in my brain after a year, but I'm open to reviewing a PR.

**Comment by Ben3eeE** [comment]
@nathansobo Yeah I started with this but then got sidetracked. I think I will open a series of PRs each removing one instance of resolving using `KeyboardEvent.code`.

I opened https://github.com/atom/atom-keymap/pull/228 now. I've had that branch around for a while. Not much to review there since it's more testing that needs to be done.

**Comment by ncsucharlie** [comment]
@Ben3eeE I see you duped issue #16350.  I'm trying to figure out the correct code to add to my init.coffee to fix the issue.  When I press <ENTER> key the keybinding resolver tells me resolves to 'space'.  What is correct addKeystrokeResolver function to fix that issue?  Or will this bug here be fixed anytime soon?  Thanks for your time.

**Comment by Ben3eeE** [comment]
@ncsucharlie Please see my reply on discuss where I helped another user https://discuss.atom.io/t/e-key-acting-as-backspace/37075/15?u=ben3eee If you still have questions I will be happy to help out in that topic or in a new topic on discuss. Feel free to @mention me there 🙂

**Comment by Arcanemagus** [comment]
Some likely helpful information for when this can _eventually_ be fixed properly in a far future Electron version:
* [Keyboard Map API](https://github.com/wicg/keyboard-map/blob/master/explainer.md)

It looks like this is being released in [Chrome 69](https://blog.chromium.org/2018/08/chrome-69-beta-av1-video-decoder-css.html).

**Comment by Ben3eeE** [comment]
Wow thanks for sharing @Arcanemagus this sounds interesting
